lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp

```diff
@@ -474,6 +476,228 @@ LogicalResult LowerAnnotationsPass::applyAnnotation(DictionaryAttr anno,
+    for (auto [index, portInfo] : modifications.portsToAdd) {
+      // Create the port.
+      newPorts.push_back({fmodule.getNumPorts(), portInfo});
+      problemIndex.push_back(index);
+    }
+    auto originalNumPorts = fmodule.getNumPorts();
+    auto portIdx = fmodule.getNumPorts();
+    fmodule.insertPorts(newPorts);
+
+    auto builder = ImplicitLocOpBuilder::atBlockEnd(UnknownLoc::get(context),
+                                                    fmodule.getBodyBlock());
+
+    // Utility function to connect a destination to a source.  Always use a
+    // ConnectOp as the widths may be uninferred.
+    auto connect = [&](Value src, Value dest) {
+      if (foldFlow(dest) == Flow::Source)
+        std::swap(src, dest);
+      // Create RefSend/RefResolve if necessary.
+      if (dest.getType().isa<RefType>() && !src.getType().isa<RefType>()) {
+        src = builder.create<RefSendOp>(src);
+      } else if (!dest.getType().isa<RefType>() &&
+                 src.getType().isa<RefType>()) {
+        src = builder.create<RefResolveOp>(src);
+      }
+      builder.create<ConnectOp>(dest, src);
+    };
+
+    // Connect each port to the value stored in the connectionMap for this
+    // wiring problem index.
+    for (auto [index, portPair] : llvm::zip(problemIndex, newPorts)) {
+      Value src = moduleModifications[fmodule].connectionMap[index];
+      assert(src && "there did not exist a driver for the port");
+      Value dest = fmodule.getArgument(portIdx++);
+      connect(src, dest);
+    }
+
+    // If a U-turn exists, this is an LCA and we need a U-turn connection.  This
+    // is the last connection made for this module.
+    for (auto [index, dest] : moduleModifications[fmodule].uturns) {
+      Value src = moduleModifications[fmodule].connectionMap[index];
+      assert(src && "there did not exist a connection for the u-turn");
+      connect(src, dest);
+    }
+
+    // Update the connectionMap of all modules for which we created a port.
+    for (auto *inst : instanceGraph.lookup(fmodule)->uses()) {
+      InstanceOp useInst = cast<InstanceOp>(inst->getInstance());
+      auto enclosingModule = useInst->getParentOfType<FModuleOp>();
+      auto clonedInst = useInst.cloneAndInsertPorts(newPorts);
+      state.instancePathCache.replaceInstance(useInst, clonedInst);
+      // When RAUW-ing, ignore the new ports that we added when replacing (they
+      // cannot have uses).
+      useInst->replaceAllUsesWith(
+          clonedInst.getResults().drop_back(newPorts.size()));
+      useInst->erase();
+      // Record information in the moduleModifications strucutre for the module
+      // _where this is instantiated_.  This is done so that when that module is
+      // visited later, there will be information available for it to find ports
+      // it needs to wire up.  If there is already an existing connection, then
+      // this is a U-turn.
+      for (auto [index, value] : llvm::enumerate(problemIndex)) {
```

Comment:
Nit, Maybe `value` -> `portIndex` makes sense here.

Reply:
I cleaned up the names of the destructured loop variables. This now uses a more exact `[newPortIdx, problemIdx]` .